Remove Tensor.h include from TensorIterator.h

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64499
* #64173
* #64172
* #64171
* #64170
* #64169
* #64168
* __->__ #64167
* #64166

Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

peterbell10 committed:
Remove Tensor.h include from TensorIterator.h

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"




[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"




[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

ezyang commented on aten/src/ATen/TensorIterator.h:
```diff
@@ -64,17 +73,40 @@ namespace internal {
 constexpr int64_t GRAIN_SIZE = 32768;
+
+// Storage for a non-owning Tensor, without needing to include Tensor.h
+class TORCH_API OpaqueTensorRef {
```
This looks a bit suspicious. If you're non-owning, why don't you just store TensorImpl*? You can reconstitute OptionalTensorRef from that without any trouble. The naming convention is also a bit suspect here.

peterbell10 replied:
>If you're non-owning, why don't you just store TensorImpl*?

`TensorIterator::tensor()`, returns a `const Tensor&` so to maintain a compatible interface we need an honest to goodness `Tensor` object that outlives the member function call. So, it needs to be stored in the `TensorIterator`. It would be almost the same to return the tensor by value, but then we need to include `Tensor.h`, defeating the purpose of this PR.

> The naming convention is also a bit suspect here.

It is "Opaque" in the same sense that you can use `Tensor&` as an opaque type, i.e. without defining the `Tensor` class. It's a "Ref" in the sense of not incrementing the reference count.

ezyang replied:
blegh ok

peterbell10 replied:
Yeah, I agree it's ugly. I really wish I could just write `reinterpret_cast<const Tensor&>(op.tensor_base)` and it not be UB.

ezyang replied:
Since you are getting an OptionalTensorRef out of this class, I'd probably prefer a wordier name like OpaqueOptionalTensorRef

ezyang replied:
meh I guess I don't care too much about this

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -43,37 +48,82 @@ inline void get_strides(int64_t* strides, ArrayRef<OperandInfo> operands, int64_
 
+static OptionalTensorRef make_otr(const TensorBase &tensor) {
+  if (tensor.defined()) {
+    return OptionalTensorRef(tensor);
```
man, it's cool that this works haha. Incredible.

ezyang commented on aten/src/ATen/TensorIterator.cpp:
```diff
@@ -414,9 +464,9 @@ void TensorIteratorBase::compute_types(const TensorIteratorConfig& config) {
+        TORCH_INTERNAL_ASSERT(op.tensor_base().defined());
         // Marker [Output original_tensor is set]
-        op.original_tensor = op.tensor;
+        op.original_tensor(op.unsafeReleaseTensor());
```
Use of unsafe is suspicious. Is there not a safe way to do this?

peterbell10 replied:
There is a safe way, but it involves refcount bumps:
```
op.original_tensor(c10::MaybeOwned<TensorBase>::owned(op.tensor_base()))
```

ezyang replied:
a small argument why it's correct would be a big help. I don't know what `original_tensor` takes as argument locally here

ezyang replied:
Actually, why don't you make `std::move(op).tensor_base()` work, which moves the tensor base out of it?

peterbell10 replied:
>Actually, why don't you make `std::move(op).tensor_base()` work, which moves the tensor base out of it?

This doesn't work because `tensor_base()` returns a reference to `TensorBase`, while `tensor_base_` is a `MaybeOwned<TensorBase>` so we would lose the borrowed status.

Looking at the again, there are only ever two things done with the unsafe ops: exchanging tensor into original_tensor, or moving original_tensor back to tensor. So instead of having the unsafe methods, I've just made these dedicated functions of `OperandInfo`.

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

ezyang committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

ezyang commented:
This one internally failing with

```
caffe2/aten/src/ATen/native/DistributionTemplates.h:167:25: error: no member named 'are_expandable' in namespace 'at'
  bool expandable = at::are_expandable(mean.sizes(), std.sizes());
                    ~~~~^
caffe2/aten/src/ATen/native/DistributionTemplates.h:171:22: error: no member named 'infer_size' in namespace 'at'
    auto shape = at::infer_size(mean.sizes(), std.sizes());
                 ~~~~^
caffe2/aten/src/ATen/native/DistributionTemplates.h:176:19: error: no member named 'resize_' in namespace 'at::native'
      at::native::resize_(output, shape);
      ~~~~~~~~~~~~^
caffe2/aten/src/ATen/native/DistributionTemplates.h:194:19: error: no member named 'resize_' in namespace 'at::native'
      at::native::resize_(output, mean.sizes());
      ~~~~~~~~~~~~^
caffe2/aten/src/ATen/native/DistributionTemplates.h:204:29: error: no member named 'view_as_real' in namespace 'at'
    auto float_tensor = at::view_as_real(self);
                        ~~~~^
```

ezyang commented:
It looks like it's pytorch/csprng/torchcsprng/csrc/cpu/kernels.cpp compilation failure

peterbell10 committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

peterbell10 commented:
Looks like a simple missing include.

ezyang committed:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

ezyang commented:
@peterbell10: @dagitses is going to take over landing this PR set

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]

dagitses commented:
@dagitses has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30728579).

Commit:
Update on "Remove Tensor.h include from TensorIterator.h"


Differential Revision: [D30728579](https://our.internmc.facebook.com/intern/diff/D30728579)

[ghstack-poisoned]
